### PR TITLE
Free trials: Add free trial exception message

### LINKED ISSUE #1482: Free Trials: Update /plans/compare/ to indicate that a domain isn't included in the trial
Since we aren't going to include domains in trials we need to be sure we indicate that a free trial of our plans does not include domains.

I suggest we add a subtle asterisk (or similar) to the /plans/compare/ page for domains and include a subtle note at the bottom of the page stating that domains are not included during the free trial period.

### LINKED ISSUE #2149: Free trials: Show free trial option on /plans/compare
We should allow users to select a free trial from /plans/compare, as well as from /plans:

screen shot 2016-01-06 at 10 20 24

Don't forget to consider how this will look in /start and when the user doesn't have a site selected.

### PR DESCRIPTION
**Pull request by scruffian**
This PR adds an exception message to the /plans/compare screen like this:

screen shot 2015-12-17 at 8 01 15 am

The message will only show in cases where the user is being offered a free trial. There are five cases to consider.
Show the message when:
- 1. A new signup is offered a free trial
- 2. An existing site is offered a free trial
- 3. A site which is in a trial already, on /plans/compare/:domain

Don't show the message when
- 4. A new signup which is not offered a free trial
- 5. An existing site is not eligible for any free trials

Fixes #1482.

This also fixes #2149 - making it possible to start a trial from /plans/compare

#### Testing instructions 1

1. Open http://calypso.localhost:3000/start/free-trial
2. Go through the flow to the plans step, and view the plans comparison screen
3. Assert that you are shown the exception message for custom domains
4. Assert that you are given the option to start a free trial

#### Testing instructions 2

1. Open http://calypso.localhost:3000/plans
2. Select a site which has never had a trial before
3. Open the plans compare screen
4. Assert that you are shown the exception message for custom domains
5. Assert that you are given the option to start a free trial

#### Testing instructions 3

1. Open http://calypso.localhost:3000/plans
2. Select any site
3. Open the plans compare screen
2. Use the site selector to select a site which is currently on a free trial
4. Assert that you are shown the exception message for custom domains

#### Testing instructions 4

1. Open http://calypso.localhost:3000/start
2. Go through the flow to the plans step, and view the plans comparison screen
3. Assert that you are not shown the exception message for custom domains
4. Assert that you are not given the option to start a free trial

#### Testing instructions 5

1. Open http://calypso.localhost:3000/plans
2. Select a site which never had a trial before
3. Open the plans compare screen
4. Assert that you are not shown the exception message for custom domains
5. Assert that you are not given the option to start a free trial

#### Reviews

- [x] Code
- [x] Product

### CODE REVIEW

File: client/components/plans/plans-compare/index.jsx [hunk 1]
```diff
@@ -74,7 +75,7 @@ var PlansCompare = React.createClass( {
 		return featuresList.map( function( feature ) {
 			return (
 				<PlanFeatureCell key={ feature.product_slug } title={ feature.description }>
-					{ feature.title }
+					{ feature.title } { featuresListUtils.featureNotPartOfTrial( feature ) ? '*' : '' }
```
Comment (stephanethomas) [inline client/components/plans/plans-compare/index.jsx]: You could also use the following shorter syntax:

```javascript
{ featuresListUtils.featureNotPartOfTrial( feature ) && '*' }
```

File: client/components/plans/plans-compare/index.jsx [hunk 2]
```diff
@@ -96,6 +99,39 @@ var PlansCompare = React.createClass( {
 		}, this );
 	},
 
+	showFreeTrialException: function() {
+		const hasTrial = this.props.sites ? this.props.sites.getSelectedSite().plan.free_trial : false,
+			canStartTrial = this.props.sitePlans ? this.props.sitePlans.data.some( function( plan ) {
```
Comment (drewblaisdell) [inline client/components/plans/plans-compare/index.jsx]: Not necessary to change, but we have used [`_.property`](https://lodash.com/docs#property) for this in the past:

```js
canStartTrial = this.props.sitePlans
  ? this.props.sitePlans.data.some( property( 'canStartTrial' ) )
  : false;
```

File: client/components/plans/plans-compare/index.jsx [hunk 3]
```diff
@@ -96,6 +99,39 @@ var PlansCompare = React.createClass( {
 		}, this );
 	},
 
+	showFreeTrialException: function() {
+		const hasTrial = this.props.sites ? this.props.sites.getSelectedSite().plan.free_trial : false,
```
Comment (drewblaisdell) [inline client/components/plans/plans-compare/index.jsx]: If `canStartTrial` returns false for sites that are currently on the free trial, I don't believe we need this condition.

File: client/lib/features-list/utils.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+function featureNotPartOfTrial( feature ) {
+	return feature.not_part_of_free_trial;
+}
+
+function featureListHasAFreeTrialException( featuresList ) {
+	return featuresList.some( function( feature ) {
```
Comment (drewblaisdell) [inline client/lib/features-list/utils.js]: You don't need to create a new anonymous function here:

```js
return featuresList.some( featureNotPartOfTrial );
```

Also, this function is only used once, and `featuresList.some( featureNotPartOfTrial );` by itself is easy to read and understand. What do you think of getting rid of this function altogether?

**Comment by fabianapsimoes** [comment]
This doesn't look super sharp:

screenshot-calypso localhost 3000 2016-01-07 15-10-37

It's too close to the rest of the copy, which makes it look kinda messy. Pinging @mikeshelton1503 for a design review. 

> 3 - A site which is in a trial already, on /plans/compare/:domain

A curiosity: how do users on trial reach `/plans/compare/:domain`? It works fine if I access that route directly, but I couldn't find a link to it.

> 5 - An existing site is not eligible for any free trials

Hmm, I still see the Free Trial button in `/plans` for existing sites.

**Comment by mikeshelton1503** [comment]
>It's too close to the rest of the copy, which makes it look kinda messy. Pinging @mikeshelton1503 for a design review.

Let's add some space above the exception message. Also it does look odd that the exception message is larger than the free trial copy so let's reduce the exception message font-size to match the font-size of the free trial copy. Lastly, to differentiate the exception message let's italicize it.

File: client/lib/features-list/utils.js (hunk 4 above)
Comment (Tug) [inline client/lib/features-list/utils.js]: Agreed :+1:

**Comment by Tug** [comment]
I took into account @drewblaisdell's feedback on the code and @mikeshelton1503's on the exception message and updated the PR.
Feel free to review those changes.

**Comment by drewblaisdell** [comment]
Code :+1:

**Comment by Tug** [comment]
This is what it looks like now:

y2ikjtsfqk-3000x3000

**Comment by Tug** [comment]
I made a last minor modification to the code to make it more consistent.

**Comment by fabianapsimoes** [comment]
I'm seeing the message when a site isn't eligible for trials:

screenshot-calypso localhost 3000 2016-01-08 11-57-17

All other cases are OK.

**Comment by scruffian** [comment]
>I'm seeing the message when a site isn't eligible for trials:

A site is ineligible for a trial when it has already had one before.

**Comment by Tug** [comment]
Ok, I think we're good now, can you review one more time @fabianapsimoes ? Thx :)

File: client/my-sites/plans/controller.jsx [hunk 5]
```diff
@@ -123,6 +123,7 @@ module.exports = {
 				<ReduxProvider store={ context.store }>
 					<CartData>
 						<PlansCompare sites={ sites }
+							enableFreeTrials={ true }
```
Comment (stephanethomas) [inline client/my-sites/plans/controller.jsx]: A quick note that you can pass the prop without value here since it's `true`.

**Comment by fabianapsimoes** [comment]
Now this case is broken:

> 3. A site which is in a trial already, on /plans/compare/:domain

We're not displaying the message when a site has an active trial. We should.

File: client/components/plans/plans-compare/index.jsx [hunk 6]
```diff
@@ -96,6 +100,53 @@ var PlansCompare = React.createClass( {
...
+		if ( this.showFreeTrialException() && featuresListUtils.featureNotPartOfTrial( feature ) ) {
+			return '*';
+		}
+
+		return null;
+	},
+
+	freeTrialExceptionMessage: function( featuresList ) {
+		if ( this.showFreeTrialException() && featuresList.some( featuresListUtils.featureNotPartOfTrial ) ) {
+			return <div className="plans-compare__free-trial-exception-message">{ this.translate( '* Not included during the free trial period' ) }</div>;
```
Comment (stephanethomas) [inline client/components/plans/plans-compare/index.jsx]: This line if already quite long. I think we could easily address that with indentation:

```
return (
  <div className="plans-compare__free-trial-exception-message">
    { this.translate( '* Not included during the free trial period' ) }
  </div>
);
```

**Comment by stephanethomas** [comment]
I added two comments but otherwise code looks good to me.

**Comment by Tug** [comment]
Thank you @fabianapsimoes for testing this over and over.
@stephanethomas I have addressed your comments ;)

**Comment by fabianapsimoes** [comment]
Looks good now. Product :+1:

**Pull request merged**